### ISSUE TITLE
[feature] allow a package to require something from a completely different profile.

### ISSUE THREAD

**Issue by smoofra** [conan-io/conan#13469]
### What is your suggestion?

Hi! I would like to be able to have a package require a different package from a completely different profile.

For example, I have a conan package for `gcc`, which builds a gcc cross compiler.   This package needs to depend on `glibc` a package for the linux C library, because it needs to copy the files from the glibc install into the sysroot for the cross compiler.     But there doesn't seem to be any way to express this dependency in conan's graph model, because glibc needs to be built with a profile for the *target* platform, not the either the -- from the gcc package's perspective -- the host platform or the build platform.

Currently I work around this by running a separate `conan install` from the `build()` method of `gcc`, but this is an unfortunate hack.

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by smoofra** [comment]
related stackoverflow question with my best attempt at shohorning this idea into conan's graph model:

https://stackoverflow.com/questions/75761045/is-it-possible-to-model-building-a-gcc-cross-compiler-in-conan-2-0

**Comment by memsharded** [comment]
Hi @smoofra 

I think there could be some approaches worth checking:

- The first one is to explicitly put ``glibc`` with the settings you want with the "per-package settings" feature. You can do in your profiles:
```
  [settings]
  compiler=gcc
  glibc/*:compiler=other
```

Same for other settings. And this can be done also in the "build" profile, which would make sense because this ``glibc`` will be in the build context. The ``profile.py`` plugin might also be useful to inject dynamically this kind of thing to all your profiles, if this is systematic
- The second approach is to do a special build in the ``glibc`` recipe, forcing the build to use the ``settings_target`` instead of the regular ``settings``, and make sure the ``package_id()`` uses the ``settings_target`` too. This one might be a bit trickier but I think it would be possible, this is typically what would be necessary to do to build it if it were part of the ``gcc`` package and built together with it (a common approach)
- The third approach could be to do a local build, with the config that you intend and then ``conan export-pkg`` it with the settings that will be used in the "build" context (same settings as the ``gcc`` gets)

The thing is that we have already consider the possibility to add new transition from the "build" context to the "host" context, but we had to discard the idea, it was introducing a ton of extra complexity, for a use case that is not that common, and that while they might require some extra hacking, the "tool_requires" are typically packages that once they are created, it is quite hidden for most usage.

> Currently I work around this by running a separate conan install from the build() method of gcc, but this is an unfortunate hack.

Yes, this is documented undefined behavior in Conan 2.0, better not do it, please check https://docs.conan.io/2/knowledge/guidelines.html

**Comment by smoofra** [comment]
oh, i like that third idea a lot.   I could build `glibc` in the normal way and then copy the files to a `sysroot` package with the same settings as `gcc`.    The CI scripts would still need to be more complicated than `conan install --build=missing` but it would eliminate the sketchy hack of calling out to conan from `build()`

**Comment by smoofra** [comment]
@memsharded 
So would it be appropriate to use things like `ConanAPI().list.select` in `package()` method of `sysroot` to find the packages that need to be copied into the sysroot, like `glibc`, `mingw-w64`  and `linux-headers` ?

The workflow would then be:
* `conan create -pr:h $TARGET glibc --build=misssing` 
* `conan export-pkg -pr $BUILD sysroot`  # copies files from glibc 
* `conan create -pr $BUILD gcc -o arch_target=$ARCH_TARGET -o os_target=$OS_TARGET` # copies them again from sysroot

**Comment by memsharded** [comment]
> So would it be appropriate to use things like ConanAPI().list.select in package() method of sysroot to find the packages that need to be copied into the sysroot, like glibc, mingw-w64 and linux-headers ?

I am afraid not, calling the API is also re-entrant, equivalent to calling Conan commands. Not possible from recipes.

But I think you want to use directly the ``self.dependencies["mydep"].package_folder`` to access the dependency folder and copy from there what you need? You also have ``self.dependencies["mydep"].cpp_info.includedirs`` etc, that will give the absolute paths.

**Comment by memsharded** [comment]
I might be missing some details to fully understand the workflow. Like the ``sysroot`` exist as a pre-compiled binary? why it is precompiled, but still requires a conan packaged ``glibc``? I also need to understand why the arch/os target options are necessary, you will have them as settings if you use:

```bash
conan create gcc -pr:b $BUILD -pr:h $TARGET --build-require
```

Probably it makes sense that we setup a quick Github project with fake recipes (not actually building C++ code, but faking it with text files) that emulates the full setup?

**Comment by smoofra** [comment]
one detail i didn't mention, sysroot actually depends on two different glibc packages, because it can be a multiarch gcc that supports both linxu-x86 and linux-x86_64.   

I'll set up the fake recipe repo

**Comment by memsharded** [comment]
Depending on 2 different architectures of the same package at the same time would be challenging. Probably name them differently if possible ``glibc32`` and ``glibc64``. Or having the sysroot also be architecture dependent?

**Comment by smoofra** [comment]
@memsharded  

https://github.com/smoofra/fake-gcc-conan

**Comment by smoofra** [comment]
I don't think naming them glibc32 and glibc64 would help.. sysroot would still need to depend on them with both sets of settings, which is impossible to do as a legitimate graph dependency in conan right?

**Comment by memsharded** [comment]
> I don't think naming them glibc32 and glibc64 would help.. sysroot would still need to depend on them with both sets of settings, which is impossible to do as a legitimate graph dependency in conan right?

yeah, seems complicated. I think our best bet could be a custom command that gathers and repackage things. I'll have a look to your project and try to think of something.

**Comment by smoofra** [comment]
thanks a lot I appreciate the help!

**Comment by memsharded** [comment]
I'll follow in the repo, opening issues to ask specific questions. We can later summarize here.

**Comment by memsharded** [comment]
I think this was also addressed by https://github.com/conan-io/conan/pull/13529, so closing the issue as solved in 2.0.5.
Please re-open or create new tickets for any further questions if necessary. Thanks!
